TerraFirmaCraft+ crashes upon game loading

### Your GTNH Discord Username

argo._

### Mod Version

1.0.0-alpha18

### Java Version

Java 8

### Bug Report

https://mclo.gs/oqaiVXg

### Mod List or GTNH Pack Version

![image](https://github.com/GTNewHorizons/Angelica/assets/96150927/7d83436d-93ce-49ff-aa24-089c806b4ff1)

And Forge 10.13.4.1614. I've tried to isolate it as much as possible.

Whilst I've seen similar issues submitted of people getting into the game and it crashing, not seen with TerraFirmaCraft+.

[CurseForge download page](https://legacy.curseforge.com/minecraft/mc-mods/terrafirmacraftplus/files/3779230) (Source code available as an attached file to each download.)

### Final Checklist

- [X] I have searched the issues and haven't found a similar issue.
- [X] I have read the known incompatibilities and this is not related to one of those.
- [X] I am running an officially released version.  (Or, if I've compiled it myself I plan to fix the issue)
- [X] This issue is not shader related  (they aren't ready yet for testing)

```
Caused by: java.lang.NoSuchFieldError: field_149760_C
	at com.dunk.tfc.Blocks.Vanilla.BlockCustomLeaves.func_149646_a(BlockCustomLeaves.java:186)
	at net.minecraft.client.renderer.RenderBlocks.func_147751_a(RenderBlocks.java:4334)
	at net.minecraft.client.renderer.RenderBlocks.func_147784_q(RenderBlocks.java:4247)
```

```
Intermediate Name	field_149760_C
Mapped Name	minY
```

--- a/src/main/java/com/gtnewhorizons/angelica/transform/BlockTransformer.java
+++ b/src/main/java/com/gtnewhorizons/angelica/transform/BlockTransformer.java
@@ -11,6 +11,7 @@ import java.util.List;
 
 public class BlockTransformer implements IClassTransformer {
     public static final String BlockClass = "net/minecraft/block/Block";
+    public static final String BlockPackage = BlockClass.substring(0, BlockClass.lastIndexOf('/') + 1);
     private static final String BlockClassFriendly = BlockClass.replace('/', '.');
     public static final List<Pair<String, String>> BlockBoundsFields = ImmutableList.of(
         Pair.of("minX", "field_149759_B"),

--- a/src/main/java/com/gtnewhorizons/angelica/transform/ClassConstantPoolParser.java
+++ b/src/main/java/com/gtnewhorizons/angelica/transform/ClassConstantPoolParser.java
@@ -61,6 +61,16 @@ public class ClassConstantPoolParser {
      * for
      */
     public boolean find(byte[] basicClass) {
+        return find(basicClass, false);
+    }
+    /**
+     * Returns true if the constant pool of the class represented by this byte array contains one of the Strings we are looking
+     * for.
+     *
+     * @param prefixes If true, it is enough for a constant pool entry to <i>start</i> with one of our Strings to count as a match -
+     * otherwise, the entire String has to match.
+     */
+    public boolean find(byte[] basicClass, boolean prefixes) {
         if (basicClass == null || basicClass.length == 0) {
             return false;
         }
@@ -92,9 +102,9 @@ public class ClassConstantPoolParser {
                     final int strLen = readUnsignedShort(index + 1, basicClass);
                     size = 3 + strLen;
                     for (byte[] bytes : BYTES_TO_SEARCH) {
-                        if (strLen == bytes.length) {
+                        if (prefixes ? strLen >= bytes.length : strLen == bytes.length) {
                             boolean found = true;
-                            for (int j = index + 3; j < index + 3 + strLen; j++) {
+                            for (int j = index + 3; j < index + 3 + bytes.length; j++) {
                                 if (basicClass[j] != bytes[j - (index + 3)]) {
                                     found = false;
                                     break;

--- a/src/main/java/com/gtnewhorizons/angelica/transform/RedirectorTransformer.java
+++ b/src/main/java/com/gtnewhorizons/angelica/transform/RedirectorTransformer.java
@@ -36,6 +36,7 @@ import java.util.stream.Collectors;
 
 import static com.gtnewhorizons.angelica.transform.BlockTransformer.BlockBoundsFields;
 import static com.gtnewhorizons.angelica.transform.BlockTransformer.BlockClass;
+import static com.gtnewhorizons.angelica.transform.BlockTransformer.BlockPackage;
 
 /**
  * This transformer redirects all Tessellator.instance field accesses to go through our TessellatorManager.
@@ -63,8 +64,16 @@ public class RedirectorTransformer implements IClassTransformer {
         "startGame", "func_71384_a",
         "initializeTextures", "func_77474_a"
     );
+    /** All classes in <tt>net.minecraft.block.*</tt> are the block subclasses save for these. */ 
+    private static final List<String> VanillaBlockExclusions = Arrays.asList(
+        "net/minecraft/block/IGrowable",
+        "net/minecraft/block/ITileEntityProvider",
+        "net/minecraft/block/BlockEventData",
+        "net/minecraft/block/BlockSourceImpl",
+        "net/minecraft/block/material/"
+    );
 
-    private static final ClassConstantPoolParser cstPoolParser = new ClassConstantPoolParser(GL11, GL13, GL14, OpenGlHelper, EXTBlendFunc, ARBMultiTexture, TessellatorClass, BlockClass,
+    private static final ClassConstantPoolParser cstPoolParser = new ClassConstantPoolParser(GL11, GL13, GL14, OpenGlHelper, EXTBlendFunc, ARBMultiTexture, TessellatorClass, BlockPackage,
         Project);
     private static final Map<String, Map<String, String>> methodRedirects = new HashMap<>();
     private static final Map<Integer, String> glCapRedirects = new HashMap<>();
@@ -76,7 +85,7 @@ public class RedirectorTransformer implements IClassTransformer {
     );
     private static int remaps = 0;
 
-    private static final Set<String> blockSubclasses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    private static final Set<String> moddedBlockSubclasses = Collections.newSetFromMap(new ConcurrentHashMap<>());
     // Block owners we *shouldn't* redirect because they shadow one of our fields
     private static final Set<String> blockOwnerExclusions = Collections.newSetFromMap(new ConcurrentHashMap<>());
 
@@ -152,7 +161,22 @@ public class RedirectorTransformer implements IClassTransformer {
         methodRedirects.put(EXTBlendFunc, RedirectMap.newMap().add("glBlendFuncSeparateEXT", "tryBlendFuncSeparate"));
         methodRedirects.put(ARBMultiTexture, RedirectMap.newMap().add("glActiveTextureARB"));
         methodRedirects.put(Project, RedirectMap.newMap().add("gluPerspective"));
-        blockSubclasses.add(BlockClass);
+    }
+    
+    private boolean isVanillaBlockSubclass(String className) {
+        if(className.startsWith(BlockTransformer.BlockPackage)) {
+            for(String exclusion : VanillaBlockExclusions) {
+                if(className.startsWith(exclusion)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+    
+    private boolean isBlockSubclass(String className) {
+        return isVanillaBlockSubclass(className) || moddedBlockSubclasses.contains(className);
     }
 
     @Override
@@ -167,7 +191,7 @@ public class RedirectorTransformer implements IClassTransformer {
             }
         }
 
-        if (!cstPoolParser.find(basicClass)) {
+        if (!cstPoolParser.find(basicClass, true)) {
             return basicClass;
         }
 
@@ -176,19 +200,17 @@ public class RedirectorTransformer implements IClassTransformer {
         cr.accept(cn, 0);
 
         // Track subclasses of Block
-        if (blockSubclasses.contains(cn.superName)) {
-            blockSubclasses.add(cn.name);
+        if (!isVanillaBlockSubclass(cn.name) && isBlockSubclass(cn.superName)) {
+            moddedBlockSubclasses.add(cn.name);
             cstPoolParser.addString(cn.name);
         }
 
         // Check if this class shadows any fields of the parent class
-        if(blockSubclasses.contains(cn.name)) {
+        if(moddedBlockSubclasses.contains(cn.name)) {
             // If a superclass shadows, then so do we, because JVM will resolve a reference on our class to that
             // superclass
             boolean doWeShadow;
-            if(cn.name.equals(BlockClass)) {
-                doWeShadow = false; // by definition
-            } else if(blockOwnerExclusions.contains(cn.superName)) {
+            if(blockOwnerExclusions.contains(cn.superName)) {
                 doWeShadow = true;
             } else {
                 // Check if we declare any known field names
@@ -274,7 +296,7 @@ public class RedirectorTransformer implements IClassTransformer {
                     }
                 }
                 else if ((node.getOpcode() == Opcodes.GETFIELD || node.getOpcode() == Opcodes.PUTFIELD) && node instanceof FieldInsnNode fNode) {
-                    if(!blockOwnerExclusions.contains(fNode.owner) && blockSubclasses.contains(fNode.owner) && AngelicaConfig.enableSodium) {
+                    if(!blockOwnerExclusions.contains(fNode.owner) && isBlockSubclass(fNode.owner) && AngelicaConfig.enableSodium) {
                         Pair<String, String> fieldToRedirect = null;
                         for(Pair<String, String> blockPairs : BlockBoundsFields) {
                             if(fNode.name.equals(blockPairs.getLeft()) || fNode.name.equals(blockPairs.getRight())) {

